edb/language_server/main.py

```diff
@@ -126,22 +134,140 @@ def document_updated(ls: ls_server.GelLanguageServer, doc_uri: str):
             ql_ast = ql_ast_res.ok
 
             if isinstance(ql_ast, list):
-                diagnostic_set = ls_server.compile(ls, document, ql_ast)
+                diagnostic_set, _ = ls_server.compile(ls, document, ql_ast)
             else:
                 # SDL in query files?
-                ls.publish_diagnostics(document.uri, [], document.version)
                 diagnostic_set = ls_server.DiagnosticsSet()
         else:
             ls.show_message_log(f'Unknown file type: {doc_uri}')
+            diagnostic_set = ls_server.DiagnosticsSet()
             # doc_uri in ('gel.toml')
 
+        diagnostic_set.extend(document, [])  # make sure we publish for document
         for doc, diags in diagnostic_set.by_doc.items():
             ls.publish_diagnostics(doc.uri, diags, doc.version)
     except BaseException as e:
         send_internal_error(ls, e)
         ls.publish_diagnostics(document.uri, [], document.version)
 
 
+def document_definition(
+    ls: ls_server.GelLanguageServer, params: lsp_types.DefinitionParams
+) -> lsp_types.Location | None:
+    doc_uri = params.text_document.uri
+    document = ls.workspace.get_text_document(doc_uri)
+
+    position: int = qltokenizer.line_col_to_source_point(
+        document.source, params.position.line, params.position.character
+    ).offset
+    ls.show_message_log(f'position = {position}')
+
+    try:
+        if is_schema_file(doc_uri):
+            ls.show_message_log(
+                'Definition in schema files are not supported yet'
+            )
+
+        elif is_edgeql_file(doc_uri):
+
+            ql_ast_res = ls_parsing.parse(document, ls)
+            if not ql_ast_res.ok:
+                return None
+            ql_ast = ql_ast_res.ok
+
+            if isinstance(ql_ast, list):
+
+                # <DEBUG>
+                from edb.common import markup
+                with open('ql_stmt.txt', 'w') as file:
+                    markup.dump(ql_ast, file=file)
+                # </DEBUG>
+
+                _, ir_stmts = ls_server.compile(ls, document, ql_ast)
+
+                ir_stmt: irast.Statement
+                for ir_stmt in ir_stmts:
+
+                    # <DEBUG>
+                    from edb.common import markup
+                    with open('ir_stmt.txt', 'w') as file:
+                        markup.dump(ir_stmt, file=file)
+                    # </DEBUG>
+
+                    node = edb_span.find_by_source_position(ir_stmt, position)
+                    if not node:
+                        continue
+                    assert isinstance(node, irast.Base), node
+
+                    ls.show_message_log(f'node: {str(node)}')
+                    ls.show_message_log(f'span: {str(node.span)}')
+
+                    if not isinstance(node, irast.Set):
+                        return None
+                    assert ir_stmt.schema
+
+                    # if this is a ptr, find it in the schema
```

Comment:
My inclination would be to check if `expr` is a `Pointer`, and use the `ptrref` there instead. Not sure that it really matters, though.